### PR TITLE
Changes support target natural id.

### PR DESCRIPTION
**Pull request by agorina**
This is related to https://github.com/SmarterApp/RDW_Schema/pull/73.
What is left:
1. Add support for 'other' targets to the package ingest.
2. Add support for common core standards to the package ingest.
3. Modify migrate to include the above codes.

### CODE REVIEW

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/AssessmentParser.java [hunk 1]
```diff
@@ -9,6 +9,12 @@
...
-    Assessment parse(final CSVRecord record, final DataElementErrorCollector errorCollector);
-
+    /**
+     * Parse function to validate and extract an {@link Assessment}. Any errors found are gathered in a central error collector.
+     *
+     * @param record         the {@link CSVRecord} to parse
+     * @param errorCollector error collector
+     * @return {@link Assessment} if successfully validated and parsed, null otherwise
+     */
+    Assessment parse(CSVRecord record, DataElementErrorCollector errorCollector);
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/AssessmentParser.java]: Just moved a comment from the implementation to the interface, no real change here.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParser.java [hunk 2]
```diff
@@ -80,6 +69,7 @@ public Assessment parse(final CSVRecord record, final DataElementErrorCollector
                 .typeId(parserHelper.repositoryValidation(ASSESSMENT_TYPE, record.get(ASSESSMENT_TYPE).trim(),
                         this::toTypeCodeOrNull, typeRepository::findIdByCode));
 
+        //TODO: this is not exactly right, ICA must have scores while IAB does not. Need to enforce the validation
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParser.java]: I will create a JIRA to fix this.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultItemParser.java [hunk 3]
```diff
@@ -34,44 +31,31 @@
...
-     * @param errorCollector error collector
-     * @return {@link Item} if successfully validated and parsed, null otherwise
-     */
     @Override
     public Item parse(final CSVRecord record, final DataElementErrorCollector errorCollector) {
         final ParserHelper parserHelper = new ParserHelper(errorCollector);
 
-        //TODO: add more tests
-        parserHelper.repositoryValidation(ASSESSMENT_SUBJECT, record.get(ASSESSMENT_SUBJECT).trim(), Strings::emptyToNull, subjectRepository::findIdByName);
-
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultItemParser.java]: The subject is validated when the assessment is parsed. This is a duplicate check.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultItemParser.java [hunk 4]
```diff
@@ -34,44 +31,31 @@
...
                 .assessmentNaturalId(parserHelper.basicValidation(ASSESSMENT_ID, record.get(ASSESSMENT_ID).trim(),
                         Strings::emptyToNull))
-                .claimId(parserHelper.repositoryValidation(ITEM_CLAIM, record.get(ITEM_CLAIM).trim(),
+                .claimId(parserHelper.repositoryValidation(ITEM_CLAIM, toClaim(record.get(ITEM_CLAIM_TARGET).trim()),
                         Strings::emptyToNull, claimRepository::findIdByCode))
-                .targetId(parserHelper.repositoryBiValidation(ITEM_TARGET, record.get(ITEM_TARGET).trim(), Strings::emptyToNull,
-                        ITEM_CLAIM, record.get(ITEM_CLAIM).trim(), Strings::emptyToNull, targetRepository::findIdByCodeAndClaim))
+                .targetId(parserHelper.repositoryBiValidation(ITEM_TARGET, toTargetNaturalId(record.get(ITEM_CLAIM_TARGET).trim()), Strings::emptyToNull,
+                        ITEM_CLAIM, toClaim(record.get(ITEM_CLAIM_TARGET).trim()), Strings::emptyToNull, targetRepository::findIdByNaturalIdAndClaim))
                 .difficulty(parserHelper.basicValidation(DIFFICULTY, record.get(DIFFICULTY),
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultItemParser.java]: I know I am parsing it twice, but I could not find a clean way to include it into the existing flow.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/ItemParser.java [hunk 5]
```diff
@@ -9,5 +9,12 @@
...
 
-    Item parse(final CSVRecord record, final DataElementErrorCollector errorCollector);
+    /**
+     * Parse function to validate and extract an {@link Item}. Any errors found are gathered in the error collector.
+     *
+     * @param record         the {@link CSVRecord} to parse
+     * @param errorCollector the error collector
+     * @return {@link Item} if successfully validated and parsed, null otherwise
+     */
+    Item parse(CSVRecord record, DataElementErrorCollector errorCollector);
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/ItemParser.java]: Once again, just copied from the implementation.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/PackageProcessorConfigurationTest.java [hunk 6]
```diff
@@ -55,7 +55,7 @@ public void createProcessors() throws IOException {
 
     @Test
     public void itShouldProcessAssessmentPackage() throws UnsupportedEncodingException {
-        processor.processPackage(createMessage(1, PACKAGE, MediaType.TEXT_PLAIN, "/IAB_English.items.csv"));
+        processor.processPackage(createMessage(1, PACKAGE, MediaType.TEXT_PLAIN, "/IAB_ICA_Combined.items.csv"));
         verify(packageProcessor).process(any(byte[].class), eq(1L));
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/PackageProcessorConfigurationTest.java]: I changed the test files.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/AssessmentPackageRepositoryIT.java [hunk 7]
```diff
@@ -69,7 +69,7 @@ public void setUp() {
                 .naturalId("(SBAC)SBAC-Winter-2016-2017");
 
         itemBuilder.mathPractice(-11)
-                .targetId(1111)
+                .targetId(-1111)
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/AssessmentPackageRepositoryIT.java]: Was conflicting with preloaded data.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java [hunk 8]
```diff
@@ -30,48 +31,34 @@
...
-                                .getResourceAsStream("/IAB_English.items.csv")));
+                CSVFormat.RFC4180.withFirstRecordAsHeader().parse(new InputStreamReader(this.getClass().getResourceAsStream("/IAB_ICA_Combined.bad-asmt.csv")));
 
         for (CSVRecord record : records) {
-            asmtSet.add(assessmentParser.parse(record, errorCollector));
+            final Assessment assessment = assessmentParser.parse(record, errorCollector);
+            assertThat(assessment).isNull();
         }
-        assertThat(errorCollector.isEmpty());
-        assertThat(asmtSet.size() == 9);
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java]: This was not right. It was not asserting anything.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java [hunk 9]
```diff
@@ -80,8 +67,8 @@ public void itShouldHaveNoErrorsAndReturnObjectsWhenParsingGoodMathData() throws
         for (CSVRecord record : records) {
             asmtSet.add(assessmentParser.parse(record, errorCollector));
         }
+
         assertThat(errorCollector.isEmpty());
-        assertThat(asmtSet.size() == 2476);
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java]: Ditto.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java [hunk 10]
```diff
@@ -30,48 +31,34 @@
...
         DataElementErrorCollector errorCollector = new DataElementErrorCollector();
-        HashSet<Assessment> asmtSet = new HashSet<Assessment>();
-        assertThat(errorCollector.isEmpty());
-        assertThat(asmtSet.isEmpty());
+
         final Iterable<CSVRecord> records =
-                CSVFormat.RFC4180.withFirstRecordAsHeader()
-                        .parse(new InputStreamReader(this.getClass()
-                                .getResourceAsStream("/IAB_English.items.csv")));
+                CSVFormat.RFC4180.withFirstRecordAsHeader().parse(new InputStreamReader(this.getClass().getResourceAsStream("/IAB_ICA_Combined.bad-asmt.csv")));
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java]: I changed the file to be small and include the exact failures that are being checked below.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java [hunk 11]
```diff
@@ -30,48 +31,34 @@
...
-        assertThat(asmtSet.size() == 9);
+        assertThat(errorCollector.toJson()).isEqualTo("{\"messages\":[" +
+                "{\"elementName\":\"AssessmentId\",\"value\":\"\",\"error\":\"Invalid value\"}," +
+                "{\"elementName\":\"AssessmentName\",\"value\":\"\",\"error\":\"Invalid value\"}," +
+                "{\"elementName\":\"AssessmentSubject\",\"value\":\"\",\"error\":\"Invalid value\"}," +
+                "{\"elementName\":\"AssessmentGrade\",\"value\":\"\",\"error\":\"Invalid value\"}," +
+                "{\"elementName\":\"AssessmentGrade\",\"value\":\"something\",\"error\":\"Unknown value\"}," +
+                "{\"elementName\":\"AssessmentSubtype\",\"value\":\"something\",\"error\":\"Unknown value\"}" +
+                "]}");
     }
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java]: This shows what is being checked as part of the assessment validation. It includes 'subject' and that is the reason I removed it from the item validation.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcTargetRepository.java [hunk 12]
```diff
@@ -13,28 +13,23 @@
 class JdbcTargetRepository implements TargetRepository {
     private final NamedParameterJdbcTemplate jdbcTemplate;
 
-    @Value("${sql.target.findIdByCodeAndClaim}")
+    @Value("${sql.target.findIdByNaturalIdAndClaim}")
     private String findIdByCodeAndClaimSql;
```
Comment (malaffoon) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcTargetRepository.java]: rename variable?

Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcTargetRepository.java]: Yes, thank you. 'Replace All' missed it.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java (hunk 9 above)
Comment (malaffoon) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java]: The isEmpty() assertion is also not doing anything; either `assertThat(errorCollector).isEmpty()` or `assertThat(errorCollector.isEmpty()).isTrue()`.

Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParserIT.java]: You are right. Will change.

**Pull request merged**